### ISSUE TITLE
docker_container :create fails with container-only binds property element

### ISSUE THREAD

**Issue by kkeane** [chef-cookbooks/docker#493]
I reproduced this with the head of the git repository as of right now (10/11, around noon PST), haven't tried with a released version, so this could potentially be a transient problem. I stumbled across this trying to reproduce #492 . I am using docker 1.7.1 (docker API version 1.19), on CentOS 6.7.

If you specify a container volume in the binds array without the corresponding host volume, creating the container will fail, claiming that the container already exists, even though it actually doesn't.

The semantics of specifying only a container volume without a host volume is that the volume should be placed under /var/lib/docker/volumes/xxxxx/_data (the same semantics as the VOLUME entry in a Dockerfile).

Preparation: make sure that the container "bindtest_chef" is deleted.

Details:

The following resource will always fail with the message "Conflict. The name "bindtest_chef" is already in use by container a7d892f46981. You have to delete (or rename) that container to be able to reuse that name."

    docker_container "bindtest_chef" do
      action :create
      repo "centos"
      binds [ '/home', '/tmp/bindtest:/tmp' ]
    end

The following resource will succeed (the only difference is the first entry in the binds array):

     docker_container "bindtest_chef" do
      action :create
      repo "centos"
      binds [ '/home:/home', '/tmp/bindtest:/tmp' ]
    end

**Comment by someara** [comment]
I'm trying to grok this... 

you're saying that

```
docker_container 'whatever' do
  binds [ '/home', '/tmp/bindtest:/tmp' ]
end
```

... should behave like

```
docker_container 'whatever' do
  binds [ '/home', '/tmp/bindtest:/tmp' ]
  volume [ '/home', ]
end
```

?

**Comment by kkeane** [comment]
Good point, actually. Mostly, I wanted to see if that scenario might be what triggered #492 for me, so I hadn't thought of the volume property.

That said, at the API level, this is permitted. https://docs.docker.com/reference/api/docker_remote_api_v1.19/ Not allowing it at the cookbook level is of course OK; in that case, maybe you can make this an enhancement request for better error handling? Creating the container twice is a kinda surprising outcome from such a syntax error.

     Binds – A list of volume bindings for this container. Each volume binding is a string in one of these forms:
       container_path to create a new volume for the container
       host_path:container_path to bind-mount a host path into the container
       host_path:container_path:ro to make the bind-mount read-only inside the container.

**Comment by someara** [comment]
It looks like both the Docker CLI and Docker Compose both let you specify both binds and volumes as a "volume".

... which is strange to me, because 
binds means "mount this directory off the host and make it visible from inside the container
volumes means "make this directory inside the container mountable from other containers"

which seem like two completely different things to me.

I'm not sure where the current "volumes" semantics came from... if I had to guess, it's a historical artifact from The Olden Times.

It's strange to me that there would have the two endpoints if they didn't mean for people to use them... perhaps the documentation is out of date?

Chef is just passing blindly passing the string into the api here:

https://github.com/bflad/chef-docker/blob/3ef0a456e0420e817522bc02670dfb8e18a53c84/libraries/docker_container.rb#L198

so it *should* be working if the API does indeed support it.

If not, we should add some input sanity checking.

-s

**Comment by kkeane** [comment]
> binds means "mount this directory off the host and make it visible from inside the container
> volumes means "make this directory inside the container mountable from other containers"

I think the semantics of volume is actually different in a subtle way that may explain this seemingly surprising behavior. volume actually simply is a special case of bind-mounting.

bind indeed clearly is an import operation into the container.

Volume at first glance looks like an export, but it actually is an import, as well. Docker decides where the host directory goes. It looks like an export because if the volume is newly created, Docker will initialize it by *copying* the content of the image directory into the host directory.

That's actually a big pitfall: when you update the image with new files, and the old volume still exists, the volume will contain the old content, not the updated files in the image. Tripped me up a couple times when I first learned Docker.

**Comment by kkeane** [comment]
> Chef is just blindly passing the string into the api here:
> https://github.com/bflad/chef-docker/blob/3ef0a456e0420e817522bc02670dfb8e18a53c84/libraries/docker_container.rb#L198
> so it should be working if the API does indeed support it.

Here is the relevant section of the docker log file. My conclusion is that the documentation for the API is wrong AND that there is a bug in the docker API. So I agree, sanity checking the input is probably the best approach. Thanks for your patience!

Most interesting is the end. It looks like this might be a bug in the docker API, rather than your resource. The API really does reject the "invalid volume specification" but then creates the container anyway (without volumes; I checked). My guess is that the error message occurs because a container that shouldn't exist does so anyway.

It also includes a few additional relevant steps:

- My complete test script will attempt to delete the containers bindtest_script and bindtest_chef
- I then use the docker CLI to create a container named bindtest_script, matching the bindtest_chef version. I use two --volume arguments instead of the bind property.

time="2015-10-11T13:57:30.444386388-07:00" level=info msg="POST /v1.19/images/create?fromImage=centos%3Alatest" 
time="2015-10-11T13:57:34.831332384-07:00" level=info msg="DELETE /v1.19/containers/bindtest_script?force=1&v=1" 
time="2015-10-11T13:57:36.930620236-07:00" level=info msg="DELETE /v1.19/containers/bindtest_chef?force=1&v=1" 
time="2015-10-11T13:57:36.930803687-07:00" level=error msg="Handler for DELETE /containers/{name:.*} returned error: no such id: bindtest_chef" 
time="2015-10-11T13:57:36.930829376-07:00" level=error msg="HTTP Error" err="no such id: bindtest_chef" statusCode=404 
time="2015-10-11T13:57:36.950127414-07:00" level=info msg="POST /v1.19/containers/create?name=bindtest_script" 
time="2015-10-11T13:57:37.957170892-07:00" level=info msg="GET /v1.19/containers/bindtest_script/json" 
time="2015-10-11T13:57:37.986584780-07:00" level=info msg="GET /v1.19/containers/bindtest_script/json" 
time="2015-10-11T13:57:38.016024773-07:00" level=info msg="GET /v1.19/containers/json?all=1" 
time="2015-10-11T13:57:59.414168711-07:00" level=info msg="POST /v1.19/images/create?fromImage=centos%3Alatest" 
time="2015-10-11T13:58:02.313932698-07:00" level=info msg="DELETE /v1.19/containers/bindtest_script?force=1&v=1" 
time="2015-10-11T13:58:04.824344184-07:00" level=info msg="DELETE /v1.19/containers/bindtest_chef?force=1&v=1" 
time="2015-10-11T13:58:04.824520375-07:00" level=error msg="Handler for DELETE /containers/{name:.*} returned error: no such id: bindtest_chef" 
time="2015-10-11T13:58:04.824547321-07:00" level=error msg="HTTP Error" err="no such id: bindtest_chef" statusCode=404 
time="2015-10-11T13:58:04.844712632-07:00" level=info msg="POST /v1.19/containers/create?name=bindtest_script" 
time="2015-10-11T13:58:05.942824091-07:00" level=info msg="GET /v1.19/containers/bindtest_script/json" 
time="2015-10-11T13:58:05.961183652-07:00" level=info msg="GET /v1.19/containers/bindtest_script/json" 
time="2015-10-11T13:58:05.983148084-07:00" level=info msg="GET /v1.19/containers/json?all=1" 
time="2015-10-11T13:58:09.984629375-07:00" level=info msg="GET /v1.16/containers/bindtest_chef/json" 
time="2015-10-11T13:58:09.984818480-07:00" level=error msg="Handler for GET /containers/{name:.*}/json returned error: no such id: bindtest_chef" 
time="2015-10-11T13:58:09.984843750-07:00" level=error msg="HTTP Error" err="no such id: bindtest_chef" statusCode=404 
time="2015-10-11T13:58:09.985714374-07:00" level=info msg="GET /v1.16/containers/bindtest_chef/json" 
time="2015-10-11T13:58:09.985799887-07:00" level=error msg="Handler for GET /containers/{name:.*}/json returned error: no such id: bindtest_chef" 
time="2015-10-11T13:58:09.985824389-07:00" level=error msg="HTTP Error" err="no such id: bindtest_chef" statusCode=404 
time="2015-10-11T13:58:09.986544162-07:00" level=info msg="GET /v1.16/containers/bindtest_chef/json" 
time="2015-10-11T13:58:09.986622974-07:00" level=error msg="Handler for GET /containers/{name:.*}/json returned error: no such id: bindtest_chef" 
time="2015-10-11T13:58:09.986644893-07:00" level=error msg="HTTP Error" err="no such id: bindtest_chef" statusCode=404 
time="2015-10-11T13:58:09.987426864-07:00" level=info msg="GET /v1.16/containers/bindtest_chef/json" 
time="2015-10-11T13:58:09.987508608-07:00" level=error msg="Handler for GET /containers/{name:.*}/json returned error: no such id: bindtest_chef" 
time="2015-10-11T13:58:09.987530387-07:00" level=error msg="HTTP Error" err="no such id: bindtest_chef" statusCode=404 
time="2015-10-11T13:58:09.988441500-07:00" level=info msg="GET /v1.16/version" 
time="2015-10-11T13:58:09.994861520-07:00" level=info msg="POST /v1.16/containers/create?name=bindtest_chef" 
time="2015-10-11T13:58:10.885208665-07:00" level=error msg="Handler for POST /containers/create returned error: Invalid volume specification: /home" 
time="2015-10-11T13:58:10.885261229-07:00" level=error msg="HTTP Error" err="Invalid volume specification: /home" statusCode=500 
time="2015-10-11T13:58:10.888993917-07:00" level=info msg="POST /v1.16/containers/create?name=bindtest_chef" 
time="2015-10-11T13:58:10.890522396-07:00" level=error msg="Handler for POST /containers/create returned error: Conflict. The name \"bindtest_chef\" i

**Comment by kkeane** [comment]
BTW, when sanity-checking, there are other edge cases. Basically, as far as I can tell, the valid cases are where either the string contains a single : or it contains two colons and ends with :ro . I'm not sure if a colon within a path (such as, in an NFS path) is supported by Docker, but I doubt it.

**Comment by someara** [comment]
>I think the semantics of volume is actually different in a subtle way that may explain this seemingly >surprising behavior. volume actually simply is a special case of bind-mounting. Or rather - bind->mounting is a generalization of volumes.

Well how about that. I learned something today. Thank you!

I did notice some confusion when splitting volumes and binds into two separate properties. Now I know why. The current situation definitely violates the principle of least surprise.

I wonder if "binds" and "volumes" should just be aliases of each other, and both provide the expected behavior to match Docker CLI and Compose. The only reason I'm not %100 sure is because of the two fields in the API.

I obviously want to do The Right Thing here, I'm just not sure what it is. Docker is undergoing a lot of rapid changes in these areas, and it would be nice to future-proof the interface if possible.

> Here is the relevant section of the docker log file. My conclusion is that the documentation for the API is > wrong AND that there is a bug in the docker API. So I agree, sanity checking the input is probably the 
> best approach.

Can you check to see if this was fixed in 1.8? I seem to recall a very short period between 1.6 and 1.8. If this only exists in 1.7, then we can probably get away with a very simple sanity check.

> BTW, when sanity-checking, there are other edge cases. Basically, as far as I can tell, the valid cases > are where either the string contains a single : or it contains two colons and ends with :ro . I'm not sure if > a colon within a path (such as, in an NFS path) is supported by Docker, but I doubt it.

Luckily the coerce feature in the new resource model will make this very easy to implement.

Thanks Kevin!

**Comment by kkeane** [comment]
> I wonder if "binds" and "volumes" should just be aliases of each other, and both provide the expected behavior to match Docker CLI and Compose. The only reason I'm not %100 sure is because of the two fields in the API.

In the CLI, both are already covered with the same parameter, -v. But given that the API does not seem to like the single-directory version, it may not be as simple as it seems.

> Can you check to see if this was fixed in 1.8? I seem to recall a very short period between 1.6 and 1.8. If this only exists in 1.7, then we can probably get away with a very simple sanity check.

Unfortunately, I can't, at least not easily. Until I upgrade my system to CentOS 7 (which is a ways off due to other software), 1.7.1 is the latest version available on CentOS 6.7. For the same reason, the 1.7 version will remain important for a while even if the time from 1.6 to 1.8 was very short.

Thanks for all your patience and effort nailing this one down!

**Comment by chasebolt** [comment]
Alright, got this figured out.

Docker CLI parses the -v option and if there is no : then it will pass the arg to Volumes. If there is a : then the arg will be passed to Binds.

https://gist.github.com/chasebolt/2e18676f441f7c42a69b
https://gist.github.com/chasebolt/6eaaff1bf571929a71cc

Easy solution is to mimic the CLI and parse the Binds property to set Volumes.

**Comment by chasebolt** [comment]
I have this mostly working but need to finish testing and submit PR when I'm back.

**Comment by someara** [comment]
This was resolved in #607 
Fix released in 2.4.2.

Closing!
